src/lib.rs

```diff
@@ -1,334 +1,201 @@
-            ("macos", "x86_64") => "protoc-osx-x86_64",
-            ("windows", _) => "protoc-win32.exe",
-            _ => return "protoc".to_owned(),
-        };
-        let bin_path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
-            .join("bin")
-            .join(protoc_bin_name);
-        format!("{}", bin_path.display())
+    pub fn generate(&self) {
+        assert!(!self.files.is_empty(), "No files specified for generation");
+        self.prep_out_dir();
+        self.generate_files();
+        self.generate_mod_file();
     }
 
-    /// Check that the user's installed version of the protobuf compiler is 3.1.x.
-    pub fn check_protoc_version(protoc: &str) {
-        let ver_re = Regex::new(r"([0-9]+)\.([0-9]+)\.[0-9]").unwrap();
-        let ver = Command::new(protoc)
-            .arg("--version")
-            .output()
-            .expect("Program `protoc` not installed (is it in PATH?).");
-        let caps = ver_re.captures(from_utf8(&ver.stdout).unwrap()).unwrap();
-        let major = caps.get(1).unwrap().as_str().parse::<i16>().unwrap();
-        let minor = caps.get(2).unwrap().as_str().parse::<i16>().unwrap();
-        if major == 3 && minor < 1 || major < 3 {
-            panic!(
-                "Invalid version of protoc (required at least 3.1.x, get {}.{}.x).",
-                major, minor,
-            );
-        }
+    pub fn wrapper_options(&mut self, wrapper_opts: GenOpt) -> &mut Self {
+        self.wrapper_opts = wrapper_opts;
+        self
     }
-    pub fn generate<T: AsRef<Path> + Debug>(includes: &[T], files: &[T], out_dir: &str) {
-        check_protoc_version(&get_protoc());
-        let mut cmd = Command::new(get_protoc());
-        let desc_file = format!("{}/mod.desc", out_dir);
-        for i in includes {
-            cmd.arg(format!("-I{}", i.as_ref().display()));
-        }
-        cmd.arg("--include_imports")
-            .arg("--include_source_info")
-            .arg("-o")
-            .arg(&desc_file);
-        for f in files {
-            cmd.arg(&format!("{}", f.as_ref().display()));
-        }
-        println!("executing {:?}", cmd);
-        match cmd.status() {
-            Ok(e) if e.success() => {}
-            e => panic!("failed to generate descriptor set files: {:?}", e),
-        }
 
-        let desc_bytes = std::fs::read(&desc_file).unwrap();
-        let desc: protobuf::descriptor::FileDescriptorSet =
-            protobuf::parse_from_bytes(&desc_bytes).unwrap();
-        let mut files_to_generate = Vec::new();
-        'outer: for file in files {
-            for include in includes {
-                if let Ok(truncated) = file.as_ref().strip_prefix(include) {
-                    files_to_generate.push(format!("{}", truncated.display()));
-                    continue 'outer;
+    /// Finds proto files to operate on in the `proto_dir` directory.
+    pub fn search_dir_for_protos(&mut self, proto_dir: &str) -> &mut Self {
```

Comment:
Should it also be added to include?

Reply:
I'm not sure what you mean: you want to search the include directory for protos or add a search for the include directory. Generally for includes we pass a directory to the compiler, not a list of files, so I think it is not needed.

Reply:
What I mean is the params passed to protoc like
```
protoc -I include1 -I include2 file1 file2 file3
```

file1 should start with one of include1 and include2.

Reply:
OK, I think it's fine to leave it to caller.